File: sstvae/modem/ofdm.py
"""DFT-matrix OFDM: complex carrier amplitudes <-> waveform samples.

Passband modulation generates the real transmit waveform directly.
Demodulation operates on the complex baseband signal produced by
dsp.to_baseband, where carrier k sits at bin (k - 11) * RS Hz.
"""

import numpy as np

from ..config import (
    FS, RS, NC, M, NCP, NSYM, CARRIER0, FCENTER, PILOT_QUADRANTS,
)

CARRIER_FREQS = CARRIER0 + RS * np.arange(NC)  # passband, Hz
BASEBAND_FREQS = CARRIER_FREQS - FCENTER  # multiples of RS

# Every frequency here is an integer number of Hz and every sample index
# is an integer, so `n*f/FS` has an exact integer remainder and the
# phasor depends only on `(n*f) mod FS`. Reducing first, in integer
# arithmetic, is exact and keeps the argument to exp() under one turn.
#
# Without it |theta| reaches 262 rad, where one ulp is 5.7e-14, so the
# phasors carried ~3e-14 of error and *which* entries rounded which way
# was an accident of numpy's complex-array arithmetic. Two reasons that
# was worth fixing, neither of them the accuracy itself:
#
#   * sin/cos of a large argument disagree between libms -- and between
#     x86-64 and Apple silicon -- by far more than they do near zero,
#     because implementations differ in how far they carry argument
#     reduction. This made the tables non-reproducible across platforms.
#   * The C++ port checks itself against these values, and a tolerance
#     sized by the reference's error rather than the port's is a much
#     weaker statement. See docs/native-app.md.
def _phasor(cycles_num: np.ndarray, sign: int = 1) -> np.ndarray:
    """exp(sign * 2j*pi * cycles_num / FS) for integer `cycles_num`."""
    return np.exp(sign * 2j * np.pi * (np.asarray(cycles_num) % FS) / FS)


# Passband modulation matrix: symbol samples n=0..NSYM-1, phase reference
# at the start of the useful part (n=NCP). Carriers are multiples of RS,
# so the first NCP samples are a true cyclic prefix.
_n_sym = np.arange(NSYM) - NCP
MOD_MATRIX = _phasor(np.outer(_n_sym, CARRIER_FREQS))  # (NSYM, NC)

# Baseband demod matrix over one useful window (M samples).
_n_use = np.arange(M)
DEMOD_MATRIX = _phasor(np.outer(BASEBAND_FREQS, _n_use), -1)  # (NC, M)


def modulate_symbols(symbols: np.ndarray) -> np.ndarray:
    """(n_sym, NC) complex -> real waveform (n_sym * NSYM,)."""
    x = np.real(MOD_MATRIX @ symbols.T)  # (NSYM, n_sym)
    return x.T.reshape(-1)


def demod_window(z: np.ndarray, start: int, backoff: int = 0) -> np.ndarray:
    """Demodulate one useful window of baseband signal starting at `start`
    (nominal index of the first useful sample). `backoff` shifts the window
    earlier into the cyclic prefix; the resulting linear phase slope is
    absorbed by pilot equalization as long as it is applied consistently.
    Factor 2 undoes the amplitude halving of the real->analytic conversion.
    """
    s = start - backoff
    win = z[s : s + M]
    if len(win) < M:
        win = np.pad(win, (0, M - len(win)))
    return (2.0 / M) * (DEMOD_MATRIX @ win)


def pilot_sequence() -> np.ndarray:
    """Fixed unit-magnitude QPSK sequence used for preamble and frame pilots.

    Built from the frozen `config.PILOT_QUADRANTS` rather than re-drawn
    from `np.random.default_rng(PILOT_SEED)`. This sequence is part of
    the on-air format: if a future numpy changed its generator stream,
    the right behaviour is to keep transmitting the same pilots, not to
    follow numpy. See the note in config.py.
    """
    phases = np.pi / 4 + np.pi / 2 * np.asarray(PILOT_QUADRANTS)
    return np.exp(1j * phases)


def preamble_waveform() -> np.ndarray:
    """Real passband preamble: pilot symbol, periodic with M over the whole
    block (double-length CP + two periods)."""
    from ..config import PREAMBLE_CP, PREAMBLE_SAMPLES

    p = pilot_sequence()
    n = np.arange(PREAMBLE_SAMPLES) - PREAMBLE_CP
    e = _phasor(np.outer(n, CARRIER_FREQS))
    return np.real(e @ p)


def preamble_template() -> np.ndarray:
    """Complex baseband replica of the preamble (for timing correlation)."""
    from ..config import PREAMBLE_CP, PREAMBLE_SAMPLES

    p = pilot_sequence()
    n = np.arange(PREAMBLE_SAMPLES) - PREAMBLE_CP
    e = _phasor(np.outer(n, BASEBAND_FREQS))
    return 0.5 * (e @ p)


def pilot_template() -> np.ndarray:
    """Complex baseband replica of one bare frame-pilot symbol's useful
    window (no CP) — used by sync.acquire_blind() to find frame timing
    purely from the pilot's own per-frame periodicity, without needing
    the (non-repeating) transmission-start preamble."""
    p = pilot_sequence()
    n = np.arange(M)
    e = _phasor(np.outer(n, BASEBAND_FREQS))
    return 0.5 * (e @ p)

File: native/core/ofdm/ofdm.cpp
#include "ofdm/ofdm.hpp"

#include <algorithm>
#include <cmath>
#include <numbers>
#include <stdexcept>

namespace sstvae::ofdm {
namespace {

using config::CARRIER0;
using config::FCENTER;
using config::FS;
using config::NCP;
using config::PREAMBLE_CP;
using config::PREAMBLE_SAMPLES;
using config::RS;

constexpr double PI = std::numbers::pi;
constexpr double TWO_PI = 2.0 * PI;

// exp(2i*pi*n*f/FS), with the argument reduced to one turn *exactly*,
// in integer arithmetic, before any floating-point work happens.
//
// Every frequency here is an integer number of Hz and every sample
// index is an integer, so n*f/FS has an exact integer remainder and the
// phasor depends only on (n*f mod FS)/FS. Two things follow, and both
// matter more than they look:
//
// * **Accuracy.** Computing the unreduced argument gives |theta| up to
//   262 rad, where one ulp is 5.7e-14 -- so the phasor inherits ~1e-13
//   of error from the argument alone. The reference has exactly this
//   problem: sstvae/modem/ofdm.py's MOD_MATRIX is correctly rounded for
//   some entries and a ulp out for others, and the resulting phasors
//   differ from the true value by up to ~3e-14. Reduced, |theta| < 2*pi
//   and one ulp is 8.9e-16.
//
// * **Cross-platform determinism**, which is the real reason. sin/cos
//   of 262 rad disagree between glibc, musl and MSVC by far more than
//   they do near zero, because they differ in how far they carry
//   argument reduction. Under 2*pi they agree to well under an ulp. The
//   parity tolerance therefore has to hold on three platforms, and this
//   is what makes that a safe promise rather than a hope.
//
// The residual disagreement with Python is bounded by Python's own
// large-argument error, not by ours -- see PHASOR_TOL in the tests.
cdouble carrier_phasor(std::int64_t n, std::int64_t f) {
    const std::int64_t num = n * f;
    // Euclidean remainder: n is negative for the cyclic prefix.
    std::int64_t q = num % FS;
    if (q < 0) q += FS;
    return std::polar(1.0, TWO_PI * (static_cast<double>(q) / static_cast<double>(FS)));
}

struct Matrices {
    // Frequencies are held as integers as well as doubles: the integer
    // copies are what carrier_phasor needs for exact argument
    // reduction, and the doubles are the public API's shape.
    std::array<std::int64_t, NC> carrier_hz{};
    std::array<std::int64_t, NC> baseband_hz{};
    std::array<double, NC> carrier{};
    std::array<double, NC> baseband{};
    std::vector<cdouble> mod;    // (NSYM, NC)
    std::vector<cdouble> demod;  // (NC, M)
    std::array<cdouble, NC> pilot{};

    Matrices()
        : mod(static_cast<std::size_t>(NSYM) * NC),
          demod(static_cast<std::size_t>(NC) * M) {
        for (int k = 0; k < NC; ++k) {
            const std::size_t u = static_cast<std::size_t>(k);
            carrier_hz[u] = CARRIER0 + static_cast<std::int64_t>(RS) * k;
            baseband_hz[u] = carrier_hz[u] - FCENTER;
            carrier[u] = static_cast<double>(carrier_hz[u]);
            baseband[u] = static_cast<double>(baseband_hz[u]);
        }

        for (int n = 0; n < NSYM; ++n) {
            for (int k = 0; k < NC; ++k)
                mod[static_cast<std::size_t>(n) * NC + static_cast<std::size_t>(k)] =
                    carrier_phasor(n - NCP, carrier_hz[static_cast<std::size_t>(k)]);
        }

        for (int k = 0; k < NC; ++k) {
            for (int n = 0; n < M; ++n)
                demod[static_cast<std::size_t>(k) * M + static_cast<std::size_t>(n)] =
                    carrier_phasor(-n, baseband_hz[static_cast<std::size_t>(k)]);
        }

        // Zadoff-Chu as an exact rational turn: phase = 2*pi * NUM[k] / DEN.
        // Not the closed form -pi*u*k^2/NC -- that argument reaches -69 rad,
        // where sin/cos differ between libms, and the reference reduces for
        // exactly that reason. Same expression, same values, both sides.
        for (int k = 0; k < NC; ++k) {
            const double phase =
                2.0 * PI *
                static_cast<double>(config::PILOT_PHASE_NUM[static_cast<std::size_t>(k)]) /
                static_cast<double>(config::PILOT_PHASE_DEN);
            pilot[static_cast<std::size_t>(k)] = std::polar(1.0, phase);
        }
    }
};

const Matrices& tables() {
    static const Matrices m;
    return m;
}

// The preamble and pilot replicas differ only in their frequency set,
// their sample count and an overall scale, so they share one builder
// rather than three near-copies that could drift apart.
std::vector<cdouble> replica(const std::array<std::int64_t, NC>& freqs, int n_samples,
                             int origin, double scale) {
    const auto& p = tables().pilot;
    std::vector<cdouble> out(static_cast<std::size_t>(n_samples));
    for (int i = 0; i < n_samples; ++i) {
        cdouble acc{0.0, 0.0};
        for (int k = 0; k < NC; ++k)
            acc += carrier_phasor(i - origin, freqs[static_cast<std::size_t>(k)]) *
                   p[static_cast<std::size_t>(k)];
        out[static_cast<std::size_t>(i)] = scale * acc;
    }
    return out;
}

}  // namespace

const std::array<double, NC>& carrier_freqs() { return tables().carrier; }
const std::array<double, NC>& baseband_freqs() { return tables().baseband; }
std::span<const cdouble> mod_matrix() { return tables().mod; }
std::span<const cdouble> demod_matrix() { return tables().demod; }
const std::array<cdouble, NC>& pilot_sequence() { return tables().pilot; }

std::vector<double> modulate_symbols(std::span<const cdouble> symbols,
                                     std::size_t n_sym) {
    if (symbols.size() != n_sym * NC)
        throw std::invalid_argument("modulate_symbols: expected n_sym * NC symbols");
    const auto& mod = tables().mod;
    std::vector<double> out(n_sym * NSYM);
    for (std::size_t i = 0; i < n_sym; ++i) {
        const cdouble* s = symbols.data() + i * NC;
        for (int n = 0; n < NSYM; ++n) {
            const cdouble* row = mod.data() + static_cast<std::size_t>(n) * NC;
            // Accumulated as complex and only then reduced to its real
            // part, as `np.real(MOD_MATRIX @ symbols.T)` does. Summing
            // the real parts alone would be algebraically identical and
            // would round differently.
            cdouble acc{0.0, 0.0};
            for (int k = 0; k < NC; ++k) acc += row[k] * s[k];
            out[i * NSYM + static_cast<std::size_t>(n)] = acc.real();
        }
    }
    return out;
}

std::array<cdouble, NC> demod_window(std::span<const cdouble> z, std::int64_t start,
                                     std::int64_t backoff) {
    const std::int64_t s = start - backoff;
    if (s < 0)
        throw std::invalid_argument(
            "demod_window: window starts before the signal; the Python "
            "reference wraps a negative slice here and returns garbage");

    // Zero-padded copy of the useful window, so the matrix product below
    // is a plain fixed-length loop no matter where the window lands.
    std::array<cdouble, M> win{};
    const std::size_t us = static_cast<std::size_t>(s);
    const std::size_t avail = us < z.size() ? std::min<std::size_t>(M, z.size() - us) : 0;
    for (std::size_t i = 0; i < avail; ++i) win[i] = z[us + i];

    const auto& dm = tables().demod;
    std::array<cdouble, NC> out{};
    const double gain = 2.0 / static_cast<double>(M);
    for (int k = 0; k < NC; ++k) {
        const cdouble* row = dm.data() + static_cast<std::size_t>(k) * M;
        cdouble acc{0.0, 0.0};
        for (int n = 0; n < M; ++n) acc += row[n] * win[static_cast<std::size_t>(n)];
        out[static_cast<std::size_t>(k)] = gain * acc;
    }
    return out;
}

std::vector<double> preamble_waveform() {
    const std::vector<cdouble> e =
        replica(tables().carrier_hz, PREAMBLE_SAMPLES, PREAMBLE_CP, 1.0);
    std::vector<double> out(e.size());
    for (std::size_t i = 0; i < e.size(); ++i) out[i] = e[i].real();
    return out;
}

std::vector<cdouble> preamble_template() {
    return replica(tables().baseband_hz, PREAMBLE_SAMPLES, PREAMBLE_CP, 0.5);
}

std::vector<cdouble> pilot_template() {
    return replica(tables().baseband_hz, M, 0, 0.5);
}

}  // namespace sstvae::ofdm
